kmod/src/server.c

```diff
@@ -380,6 +380,79 @@ static int alloc_move_empty(struct super_block *sb,
 				  dst, src, le64_to_cpu(src->total_len));
 }
 
+static void init_trans_seq_key(struct scoutfs_key *key, u64 seq, u64 rid)
+{
+	*key = (struct scoutfs_key) {
+		.sk_zone = SCOUTFS_TRANS_SEQ_ZONE,
+		.skts_trans_seq = cpu_to_le64(seq),
+		.skts_rid = cpu_to_le64(rid),
+	};
+}
+
+static int get_last_seq(struct super_block *sb, __le64 *last_seq)
+{
+	DECLARE_SERVER_INFO(sb, server);
+	struct scoutfs_sb_info *sbi = SCOUTFS_SB(sb);
+	struct scoutfs_super_block *super = &sbi->super;
+	SCOUTFS_BTREE_ITEM_REF(iref);
+	struct scoutfs_key key;
+	__le64 tmp_last_seq;
+	int err;
+
+	down_read(&server->seq_rwsem);
+
+	init_trans_seq_key(&key, 0, 0);
+	err = scoutfs_btree_next(sb, &super->trans_seqs, &key, &iref);
+	if (err == 0) {
+		key = *iref.key;
+		scoutfs_btree_put_iref(&iref);
+		tmp_last_seq = key.skts_trans_seq;
+
+	} else if (err == -ENOENT) {
+		tmp_last_seq = super->next_trans_seq;
+	} else {
+		return err;
```

Comment:
This is returning with the server->seq_rwsem held.  I'd change it to use ret instead of err, set ret = 0 in the -enoent case.  Have the seq_rwsem protect all that put the tmp_last_seq add -1 and setting the callers last_seq outside in an if (ret == 0) block and then return ret.

Reply:
:+1: